include/recorders.h

```diff
@@ -78,49 +78,103 @@ class task_recorder {
+	std::string buffer_name;
+	bool init_from_buffer;
 	bool has_local_contribution;
-	std::optional<access_list> accesses;
-	std::optional<reduction_list> reductions;
-	std::optional<side_effect_map> side_effects;
-	command_dependency_list dependencies;
-	std::string task_name;
-	std::optional<detail::task_type> task_type;
-	std::optional<detail::collective_group_id> collective_group_id;
-	std::vector<detail::reduction_id> completed_reductions;
 
-	command_record(const abstract_command& cmd, const task& tsk, const buffer_name_map& get_buffer_debug_name);
+	explicit reduction_command_record(const reduction_command& rcmd, std::string buffer_name);
+};
+
+/// Base class for task command records
+struct task_command_record {
+	task_id tid;
+	task_type type;
+	std::string debug_name;
+	collective_group_id cgid;
+
+	explicit task_command_record(const task& tsk);
+};
+
+struct epoch_command_record : matchbox::implement_acceptor<command_record, epoch_command_record>, task_command_record {
+	epoch_action action;
+	std::vector<reduction_id> completed_reductions;
+
+	explicit epoch_command_record(const epoch_command& ecmd, const task& tsk);
+};
+
+struct horizon_command_record : matchbox::implement_acceptor<command_record, horizon_command_record>, task_command_record {
+	std::vector<reduction_id> completed_reductions;
+
+	explicit horizon_command_record(const horizon_command& hcmd, const task& tsk);
+};
+
+struct execution_command_record : matchbox::implement_acceptor<command_record, execution_command_record>, task_command_record {
+	subrange<3> execution_range;
+	bool is_reduction_initializer;
+	access_list accesses;
+	side_effect_map side_effects;
+	reduction_list reductions;
+
+	explicit execution_command_record(const execution_command& ecmd, const task& tsk, const buffer_name_map& get_buffer_debug_name);
+};
+
+struct fence_command_record : matchbox::implement_acceptor<command_record, fence_command_record>, task_command_record {
+	explicit fence_command_record(const fence_command& fcmd, const task& tsk, const buffer_name_map& get_buffer_debug_name);
+
+	access_list accesses;
+	side_effect_map side_effects;
 };
 
 class command_recorder {
   public:
-	void record(command_record&& record) { m_recorded_commands.push_back(std::move(record)); }
+	void record_command(std::unique_ptr<command_record> record) { m_recorded_commands.push_back(std::move(record)); }
 
-	const std::vector<detail::command_record>& get_commands() const { return m_recorded_commands; }
+	void record_dependency(const command_dependency_record& dependency) { m_recorded_dependencies.push_back(dependency); }
 
-	const command_record& get_command(const command_id cid) const {
-		const auto it = std::find_if(m_recorded_commands.begin(), m_recorded_commands.end(), [cid](const command_record& rec) { return rec.cid == cid; });
-		assert(it != m_recorded_commands.end());
-		return *it;
-	}
+	const std::vector<std::unique_ptr<command_record>>& get_all() const { return m_recorded_commands; }
```

Comment:
I think `get_all` is fine as-is (because it's a recorded, and to me it's obvious it returns all records), but we could do `get_records` to be more explicit.

Reply:
As you point out, I had to align the names between the `instruction_` and `command_recorder` to make it work in the generic context of `graph_query`. I don't want to introduce some kind of "getter" template just for this reason, and I'm trying to avoid the word `node` because of the overloaded meaning of the term (I also changed any mention of "node" in the former `instruction_query`'s doc-blocks to say "record" instead). I agree that the name is not ideal though.

Reply:
@PeterTh I disagree that it's obvious, because a recorder records dependencies as well.

@psalz Yeah `nodes` by itself is not a good name. What about `graph_nodes` or `vertices`?

Reply:
Alright, I went with `get_graph_nodes`.